File: solutions/4288-maximum-team-size-with-overlapping-intervals/maximum-team-size-with-overlapping-intervals.cpp
class Solution {
public:
    int maximumTeamSize(vector<int>& start, vector<int>& end) {
        int n = start.size();

        vector<int> sortedStart = start;
        vector<int> sortedEnd = end;

        sort(sortedStart.begin(), sortedStart.end());
        sort(sortedEnd.begin(), sortedEnd.end());

        int maxSize = 0;

        for (int i = 0; i < n; i++) {
            int a = upper_bound(sortedStart.begin(), sortedStart.end(), end[i]) - sortedStart.begin();
            int b = lower_bound(sortedEnd.begin(), sortedEnd.end(), start[i]) - sortedEnd.begin();

            maxSize = max(maxSize, a - b);
        }

        return maxSize;
    }
};

File: solutions/4288-maximum-team-size-with-overlapping-intervals/maximum-team-size-with-overlapping-intervals.java
class Solution {
    public int maximumTeamSize(int[] start, int[] end) {
        int n = start.length;

        int[] sortedStart = start.clone();
        int[] sortedEnd = end.clone();

        Arrays.sort(sortedStart);
        Arrays.sort(sortedEnd);

        int max = 0;

        for (int i = 0; i < n; i++) {
            int a = upperBound(sortedStart, end[i]);   // starts <= end[i]
            int b = lowerBound(sortedEnd, start[i]);   // ends < start[i]

            max = Math.max(max, a - b);
        }

        return max;
    }

    private int upperBound(int[] arr, int target) {
        int l = 0, r = arr.length;
        while (l < r) {
            int m = (l + r) / 2;
            if (arr[m] <= target) l = m + 1;
            else r = m;
        }
        return l;
    }

    private int lowerBound(int[] arr, int target) {
        int l = 0, r = arr.length;
        while (l < r) {
            int m = (l + r) / 2;
            if (arr[m] < target) l = m + 1;
            else r = m;
        }
        return l;
    }
}